### PR TITLE
Fixes for test_db

### LINKED ISSUE #21: Fix test_db.py
One error, one failure — out of 10 tests:
```
FAILED gourmet/tests/test_db.py::testUnicode::runTest - AttributeError: 'str' object has no attribute 'decode'
FAILED gourmet/tests/test_db.py::TestMoreDataStuff::test_image_data - AssertionError: 'None' != 'ÿØÿà\x00\x10JFIF\x00\x01\x01\x00\x00\x01\[2147 chars]0fÿÙ'
```
`AttributeError` might be an easy fix.
`AssertionError` is probably image related.

Medium priority.

### PR DESCRIPTION
**Pull request by takluyver**
Closes #21

All unicode related issues. :-)

### CODE REVIEW

**Comment by maweki** [comment]
As some of the methods you touched are quite small, do you mind annotating the type signatures?

File: gourmet/backends/db.py [hunk 1]
```diff
@@ -1687,20 +1684,21 @@ def get_amount_as_float (self, ing, mode=1): #1 == self.AMT_MODE_AVERAGE
...
-        else:
-            if item: item = str(item)
-            if key: key = str(key)
+        if isinstance(item, bytes): item = item.decode('utf-8', 'replace')
+        if isinstance(key, bytes): key = key.decode('utf-8', 'replace')
+        if not item or not key:
+            return
+
+        if item: item = str(item)
+        if key: key = str(key)
```
Comment (maweki) [inline gourmet/backends/db.py]: can you put the `if ... then` in seperate lines

File: gourmet/ImageExtras.py [hunk 2]
```diff
@@ -33,10 +33,9 @@ def resize_image (image, width=None, height=None):
         return image
 
 def get_image_from_string (raw):
```
Comment (maweki) [inline gourmet/ImageExtras.py]: Annotation for this would be nice

File: gourmet/ImageExtras.py [hunk 3]
```diff
@@ -50,18 +49,16 @@ def get_image_from_string (raw):
 
 def get_string_from_image (image):
```
Comment (maweki) [inline gourmet/ImageExtras.py]: Annotation for this would be nice

File: gourmet/ImageExtras.py [hunk 4]
```diff
@@ -50,18 +49,16 @@ def get_image_from_string (raw):
...
-    ofi = io.StringIO()
+    ofi = io.BytesIO()
     image = image.convert('RGB')
     image.save(ofi,"JPEG")
-    ret = ofi.getvalue()
-    ofi.close()
-    return ret
+    return ofi.getvalue()
 
 def get_string_from_pixbuf (pb):
```
Comment (maweki) [inline gourmet/ImageExtras.py]: Annotation for this would be nice

**Comment by takluyver** [comment]
Done the requested annotations & reformatting

File: gourmet/backends/db.py [hunk 5]
```diff
@@ -1158,10 +1157,8 @@ def validate_recdic (self, recdic):
                 import traceback
                 traceback.print_stack()
         for k,v in list(recdic.items()):
-            try:
-                recdic[k]=str(v.strip())
-            except:
-                pass
+            if isinstance(v, str):
+                recdic[k] = v.strip()
```
Comment (maweki) [inline gourmet/backends/db.py:1161]: We're allowed to update dictionaries in place now. So this works

```python
foo = {1: 2, 3: 4, 5: 6}
for k, v in foo.items():
        foo[k] = v+1
print(foo)
```

we can get rid of the `list` conversion.

**Review by maweki — CHANGES_REQUESTED** [review]
Besides the one thing (sorry that I didn't see that earlier), this looks great. Thank you

File: gourmet/backends/db.py (hunk 5 above)
Comment (takluyver) [reply gourmet/backends/db.py:1161]: I wasn't aware of a change in that space. The docs still say that you mustn't add or delete entries while iterating over a dict view; I think technically that means replacing existing entries is OK, but I'd rather play it safe and leave the `list()` call. The extra memory use should be tiny and momentary.

**Review by maweki — APPROVED** [review]
Okay then. Good work :)

**Comment by takluyver** [comment]
Thanks!

File: gourmet/ImageExtras.py [hunk 6]
```diff
@@ -32,11 +32,10 @@ def resize_image (image, width=None, height=None):
...
-def get_image_from_string (raw):
-    """Given raw image data, return an Image object."""
+def get_image_from_string(raw: bytes) -> Image.Image:
+    """Given raw image data (bytes), return an Image object."""
     if os.name =='posix':
-        sfi=io.StringIO()
-        sfi.write(raw)
+        sfi = io.BytesIO(raw)
         sfi.seek(0)
     else:
```
Comment (cydanil) [inline gourmet/ImageExtras.py:40]: This is tangential to the PR: Surely by now `BytesIO` also work non-posix, Windows namely, systems. `write_image_tempfile` could be removed?

**Review by cydanil — APPROVED** [review]
Thank you for these fixes!

File: gourmet/ImageExtras.py (hunk 6 above)
Comment (takluyver) [reply gourmet/ImageExtras.py:40]: I think you're right - the [docs for Image.save()](https://pillow.readthedocs.io/en/latest/reference/Image.html#PIL.Image.Image.save) don't mention platform restrictions. But I wouldn't try to change Windows-specific branches until there's some Windows CI set up (Azure Pipelines & Appveyor can both run on Windows). And I wouldn't try to do that before the tests are passing on Linux. :slightly_smiling_face:

Comment (takluyver) [reply gourmet/ImageExtras.py:40]: Looks like Github actions can do Windows testing too - I haven't used that.

Comment (cydanil) [reply gourmet/ImageExtras.py:40]: Agreed. I plan on adding Windows CI at a later stage, when the application will be more stable.

**Pull request merged**